engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinRootNode.java

```diff
@@ -52,4 +69,174 @@ public void setTypeName(QualifiedName typeName) {
+      if (is(CHECK_WARNINGS)) {
+        this.warnings = WarningsLibrary.getFactory().createDispatched(5);
+      }
+    }
+
+    private boolean is(byte what) {
+      return (flags & what) != 0;
+    }
+
+    @SuppressWarnings("unchecked")
+    public final <T> T processArgument(
+        VirtualFrame frame, Class<T> type, Object value, ArgContext context) {
+      assert value != null;
+      if (is(CHECK_ERRORS) && value instanceof DataflowError err) {
+        context.returnValue = err;
+        return null;
+      }
+      if (is(CHECK_PANIC_SENTINEL) && value instanceof PanicSentinel sentinel) {
+        throw sentinel.getPanic();
+      }
+      if (warnings != null) {
+        if (warnings.hasWarnings(value)) {
+          if (mapInsertAllNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            this.mapInsertAllNode = HashMapInsertAllNode.build();
+          }
+          if (this.mapInsertAllNode != null) {
+            try {
+              context.addWarnings(frame, mapInsertAllNode, warnings.getWarnings(value, false));
+              value = warnings.removeWarnings(value);
+            } catch (UnsupportedMessageException ex) {
+              throw raise(RuntimeException.class, ex);
+            }
+          }
+        }
+      }
+      if (is(REQUIRES_CAST)) {
+        var ctx = EnsoContext.get(this);
+        if (this.ensoType == null) {
+          CompilerDirectives.transferToInterpreterAndInvalidate();
+          var builtin = ctx.getBuiltins().getByRepresentationType(type);
+          if (builtin == null) {
+            this.ensoType = ctx.getBuiltins().any();
+          } else {
+            this.ensoType = builtin.getType();
+          }
+        }
+        var conv = executeConversion(value);
+        if (conv == null) {
+          CompilerDirectives.transferToInterpreter();
+          var err = ctx.getBuiltins().error().makeTypeError(this.ensoType, value, type.getName());
+          throw new PanicException(err, this);
+        }
+        return type.cast(conv);
+      } else {
+        return type.cast(value);
+      }
+    }
+
+    public final Object processWarnings(VirtualFrame frame, Object result, ArgContext context) {
+      assert context.warnings != null;
+      if (this.appendWarningNode == null) {
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+        this.appendWarningNode = insert(AppendWarningNode.build());
+      }
+      return appendWarningNode.executeAppend(frame, result, context.warnings);
+    }
+
+    abstract Object executeConversion(Object obj);
+
+    @Specialization
+    final Object extractMultiValue(EnsoMultiValue emv, @Cached EnsoMultiValue.CastToNode castTo) {
+      var extracted = castTo.findTypeOrNull(ensoType, emv, false, false);
+      return extracted;
+    }
+
+    @Fallback
+    final Object justReturnIt(Object obj) {
+      return obj;
+    }
+
+    public static ArgNode create(
+        boolean isSelf,
```

Comment:
minor: I think it would be better to have just a single `byte` argument, that would be or-ed flags

Reply:
I would have to expose the `IS_XYZ` constants to public then. That's so C-like! The aim was to have nicely typed facade and efficient internal storage.